presto-raptor/src/main/java/com/facebook/presto/raptor/storage/organization/ShardOrganizerUtil.java

```diff
@@ -13,49 +13,186 @@
-import java.util.Set;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.StringJoiner;
 
+import static com.facebook.presto.raptor.RaptorErrorCode.RAPTOR_ERROR;
+import static com.facebook.presto.raptor.metadata.DatabaseShardManager.maxColumn;
+import static com.facebook.presto.raptor.metadata.DatabaseShardManager.minColumn;
+import static com.facebook.presto.raptor.metadata.DatabaseShardManager.shardIndexTable;
+import static com.facebook.presto.raptor.metadata.JdbcUtil.jdbcType;
+import static com.facebook.presto.raptor.util.DatabaseUtil.metadataError;
 import static com.facebook.presto.spi.type.DateType.DATE;
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static com.google.common.collect.Iterables.partition;
+import static com.google.common.collect.Maps.uniqueIndex;
+import static io.airlift.slice.Slices.wrappedBuffer;
+import static java.util.Collections.nCopies;
+import static java.util.stream.Collectors.toList;
 
 public class ShardOrganizerUtil
 {
+    private static final Logger log = Logger.get(ShardOrganizerUtil.class);
+
     private ShardOrganizerUtil() {}
 
-    static Collection<Collection<ShardMetadata>> getShardsByDaysBuckets(Set<ShardMetadata> shardMetadata, Type type)
+    static Collection<ShardIndexInfo> toShardIndexInfo(
+            IDBI dbi,
+            MetadataDao metadataDao,
+            Table tableInfo,
+            Collection<ShardMetadata> shards,
+            boolean includeSortColumns)
     {
-        // bucket shards by the start day
-        ImmutableMultimap.Builder<Long, ShardMetadata> shardsByDays = ImmutableMultimap.builder();
+        Map<Long, ShardMetadata> shardsById = uniqueIndex(shards, ShardMetadata::getShardId);
+        long tableId = tableInfo.getTableId();
+
+        // include temporal range if present
+        String temporalColumnString = "";
+        Optional<TableColumn> temporalColumn = Optional.empty();
+        if (tableInfo.getTemporalColumnId().isPresent()) {
+            long temporalColumnId = tableInfo.getTemporalColumnId().getAsLong();
+            temporalColumn = Optional.of(metadataDao.getTableColumn(tableId, temporalColumnId));
+            temporalColumnString = String.format(",\n%s,\n%s", minColumn(temporalColumnId), maxColumn(temporalColumnId));
+        }
+
+        // include sort columns if needed
+        String sortColumnString = "";
+        List<TableColumn> sortColumns = metadataDao.listSortColumns(tableId);
+        if (includeSortColumns) {
+            sortColumnString = indexColumnNames(sortColumns);
+        }
+
+        ImmutableList.Builder<ShardIndexInfo> indexInfoBuilder = ImmutableList.builder();
+        try (Connection connection = dbi.open().getConnection()) {
+            for (List<ShardMetadata> partitionedShards : partition(shards, 1000)) {
+                String args = Joiner.on(",").join(nCopies(partitionedShards.size(), "?"));
```

Comment:
The old code was rather different. As long as we're changing it, might as well rename the variable to make the contents more clear.